### ISSUE TITLE
Expose public API for HTTP parser

### ISSUE THREAD

**Issue by parasyte** [njsmith/h11#30]
So I found this module by complete mistake while reading through your [admittedly excellent] [async-await](https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/) blog article. It was a serendipity find because I wasn't looking for it, but `h11` is exactly what I need.

We have a [very big] cluster of PyPy servers that has a complex connection pool written in-house to maintain its backend connections. TL;DR: [`requests`](http://docs.python-requests.org/en/master/) is great but its connection pooling does not do the kind of pooling we need.

To cut a long story short, the connection pool implementation suffers from the synchronous request behavior of every Python HTTP client on the planet. The two options to solve this are:

1. HTTP/1.1 pipelining
2. HTTP/2

Our backend [doesn't support HTTP/2](https://github.com/unbit/uwsgi/issues/1269) so strike that option.

So we've experimented with pipelining using [`http-parser`](https://github.com/benoitc/http-parser/), and well, it kind of works.

Then I found `h11`, and I was overjoyed! [Welp, that didn't last...](https://h11.readthedocs.io/en/latest/api.html#re-using-a-connection-keep-alive-and-pipelining).

> HTTP/1.1 also allows for a more aggressive form of connection re-use, in which a client sends multiple requests in quick succession, and then waits for the responses to stream back in order (“pipelining”). This is generally considered to have been a bad idea, because it makes things like error recovery very complicated.

But I'm not deterred. It's either pipelining or maintaining *several hundred thousand* open connections on the backends. And the latter is really horrifying.

For the short term, I'll be experimenting with the `h11` parsers by using their private references, and hopefully not running into anything unexpected. For the medium and long term, I'd like to see the parsers publicly exposed and documented for insane use cases like mine. :)

**Comment by njsmith** [comment]
h11 doesn't really... have a standalone parser? The state machine for the parser is the state machine for connection, and that's unlikely to change. I'm not entirely sure what you're planning here, but I guess I'll be interested to see what you come up with :-). It's unlikely that h11 will ever support pipelining, though.

It does occur to me that if you're looking for a gross solution, you might be able to get some distance by creating multiple `h11.Connection` objects, one per pipelined request. Use `send` on the first object to get the bytes for your first request and send those, then `send` on the second object to get the bytes for your next request, etc. Then when you get bytes back, feed them into the first object until you reach the end of the request, then use `Connection.trailing_data` to retrieve the unprocessed data and start feeding that into the next object...

The most challenging part of all this would be the error handling, e.g. if a connection object's state becomes MUST_CLOSE then you have to stop instead of continuing to the next object. But error handling is what makes pipelining super-challenging in general, so I guess you know what you're getting yourself into :-)

**Comment by parasyte** [comment]
Yeah, I'm going to have to bend the laws of physics on this one a bit to get it working, but looking through `h11._readers` and `h11._writers`, all the piece are there for a decent parser. If I can go one level of abstraction away from these low-level routines, that would be even better.

No worries on error handling. We have a greenlet responsible for brokering requests and responses. It has the smarts for retries and timeouts. It's just built on a simple asynchronous queue with some extra back pressure/high watermark stuff. The nice thing is all of this is working within a closed environment; this client only talks to one kind of server, and we have full control over the protocol on each side. The worst case we have to defend against is transient errors. Concretely, client and server are both speaking HTTP/1.1, and neither are allowed to send `Connection: close` headers. Keep-Alive is good, but we need to go deeper. :)

Thanks for the tip with multiple `h11.Connection` objects. If all else fails, this is another avenue that I can explore. And a list of objects would be roughly equivalent to the size-limited worker queue that we have now. So it's not all that gross, by comparison!

**Comment by parasyte** [comment]
Revisiting this old ticket, since the optimization with HTTP/1.1 pipelining has come up again recently. I did some manual testing on uWSGI (which is now extremely outdated, and development is moving at a snail's pace) ... unfortunately, it closes the connection after delivering the second response (regardless how many requests are waiting on the socket), so pipelining is not possible with uWSGI.

In other words, I can't even use this library for pipelining anyway. So I am going to close this, and switch focus to replacing uWSGI with something modern that supports HTTP/2.
